interzeptor.go

```diff
@@ -0,0 +1,140 @@
+// Package interzeptors constructs a gRPC interceptor to enforce gNSI services.
+package interzeptors
+
+import (
+	"context"
+
+	apb "github.com/openconfig/gnsi/acctz"
+	"github.com/sirupsen/logrus"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/interzeptor/acctz"
+	"google.golang.org/grpc/interzeptor/authz"
+	"google.golang.org/grpc/status"
+)
+
+var (
+	dlog = logrus.WithField("denvelope", "interzeptors")
+
+	zeptor *Interzeptor
+)
+
+type interzeptor interface {
+	Authorize(ctx context.Context) error
+}
+
+type accounter interface {
+	RecordFromContext(ctx context.Context) (*apb.RecordResponse, error)
+	Account(*apb.RecordResponse)
+}
+
+// Interzeptor is a gRPC interceptor that performs accounting for gNSI services as well as
+// enforcement of the services themselves.
+type Interzeptor struct {
+	acctMgr accounter
+
+	authz interzeptor
+	pathz interzeptor
+}
+
+// newInterzeptor constructs a new interceptor entity.
+func newInterzeptor(acctMgr accounter, authz interzeptor, pathz interzeptor) *Interzeptor {
+	return &Interzeptor{
+		acctMgr: acctMgr,
+		pathz:   pathz,
+		authz:   authz,
+	}
+}
+
+// UnaryInterceptor is a gNSI unary interceptor.
+func (i *Interzeptor) UnaryInterceptor(ctx context.Context, req any, info *grpc.UnaryServerInfo, handler grpc.UnaryHandler) (any, error) {
+	if err := i.interzept(ctx, info.FullMethod); err != nil {
+		return nil, err
+	}
+
+	return handler(ctx, req)
+}
+
+// StreamInterceptor is a gNSI stream interceptor
+func (i *Interzeptor) StreamInterceptor(srv any, ss grpc.ServerStream, info *grpc.StreamServerInfo, handler grpc.StreamHandler) error {
+	if err := i.interzept(ss.Context(), info.FullMethod); err != nil {
+		return err
+	}
+
+	return handler(srv, ss)
+}
+
+// interzept executes all registered interceptors in parallel. It retrieves any accounting
```

Comment:
It no longer does so in parallel, right? We decided that we wanted this to be serially executed, I think.

Reply:
That's correct - it is a stale comment.